### ISSUE TITLE
F2FS SSD over USB not booting since 6.33

### ISSUE THREAD

**Issue by Cjkeenan** [MichaIng/DietPi#3893]
#### Required Information
- DietPi version | `cat /boot/dietpi/.version`
```
root@BH-3B:~# cat /boot/dietpi/.version                                                                                                                                         
G_DIETPI_VERSION_CORE=6                                                                                                                                                         
G_DIETPI_VERSION_SUB=33                                                                                                                                                         
G_DIETPI_VERSION_RC=3                                                                                                                                                           
G_GITBRANCH='master'                                                                                                                                                            
G_GITOWNER='MichaIng'     
```
- Distro version | `echo $G_DISTRO_NAME` or `cat /etc/debian_version`
```
root@BH-3B:~# cat /etc/debian_version                                                                                                                                           
10.6                                                                                                                                                                            
root@BH-3B:~# echo $G_DISTRO_NAME                                                                                                                                               
buster 
```
- Kernel version | `uname -a`
```
root@BH-3B:~# uname -a                                                                                                                                                          
Linux BH-3B 5.4.72-v7+ #1356 SMP Thu Oct 22 13:56:54 BST 2020 armv7l GNU/Linux   
```
- SBC model | `echo $G_HW_MODEL_NAME` or (EG: RPi3)
```
root@BH-3B:~# echo $G_HW_MODEL_NAME                                                                                                                                             
RPi 3 Model B (armv7l) 
```
- Power supply used | (EG: 5V 1A RAVpower)
5V 2.4A APC Power Strip
- SDcard used | (EG: SanDisk ultra)
WD Blue SSD 240GB

#### Steps to reproduce

1. Have a F2FS SSD over USB as OS drive
2. Update to v6.33 of DietPi
3. Reboot System

#### Expected behaviour

- System should reboot and start all services without any user intervention

#### Actual behaviour

![image](https://user-images.githubusercontent.com/10240016/98436233-a0ca7600-208e-11eb-9d0a-05a7d4aac922.png)
- After pressing CTRL + D system boots fine and services start fine
- After entering root password, it says something like DietPi Globals could not load because the file path does not exist and nothing is in the /boot directory

#### Extra details

- If you need any more logs please let me know
- Only other thing I did before rebooting and after updating to 6.33 is doing a DietPi Backup to the same drive as the OS
- Also the boot time is very long, close to two minutes, is that normal?

**Comment by Joulinar** [comment]
Many thanks for your message.

It might not be related but you have an `under voltage` message at the end. Means your system ist not powered enough. Do you use an external power supply? Did you tried using different USB port already?

**Comment by Cjkeenan** [comment]
I am using a USB port on my APC power strip that says 2.4A Max Output, but I also tried via the Canakit 5V 2.5A power brick and that had no voltage warnings, but also had the same errors.

I entered the root password again and this is the specific output:
![image](https://user-images.githubusercontent.com/10240016/98437218-526ca580-2095-11eb-955b-5d564ac02126.png)

**Comment by Joulinar** [comment]
your SSD is not mounted. Therefore you have all the error messages and long boot times. Let's check following

```
dmesg -l emerg,alert,crit,err
lsblk -o name,fstype,label,size,ro,type,mountpoint
```

**Comment by Cjkeenan** [comment]
```
root@BH-3B:~# dmesg -l emerg,alert,crit,err                                                                                                                                                     
[    1.668161] vc_vchi_sm_init: failed to open VCHI service (-1)                                                                                                                                
[    1.668165] [vc_sm_connected_init]: failed to initialize shared memory service                                                                                                               
[  108.313382] vc_sm_cma_vchi_init: failed to open VCHI service (-1)                                                                                                                            
[  108.313396] [vc_sm_connected_init]: failed to initialize shared memory service                                                                                                               
[  108.563918] bcm2835_mmal_vchiq: Failed to open VCHI service connection (status=-1)                                                                                                           
[  110.473947] Under-voltage detected! (0x00050000)                                                                                                                                             
[  125.034033] Under-voltage detected! (0x00050005)                                                                                                                                             
[  141.687467] Under-voltage detected! (0x00050000)                                                                                                                                             
[  929.993908] Under-voltage detected! (0x00050005)                                                                                                                                             
[  952.873872] Under-voltage detected! (0x00050005)                                                                                                                                             
[  992.393841] Under-voltage detected! (0x00050000)  
```
and 
```
root@BH-3B:~# lsblk -o name,fstype,label,size,ro,type,mountpoint                                                                                                                                
NAME   FSTYPE LABEL   SIZE RO TYPE MOUNTPOINT                                                                                                                                                   
sda                 232.9G  0 disk                                                                                                                                                              
├─sda1 vfat   boot    256M  0 part /boot                                                                                                                                                        
└─sda2 f2fs         232.6G  0 part /  
```
These were both done after CTRL+D and the system boots as normal.

This is my `FSTAB` by the way:
```
                                                                                                                                                                                                
# Please use "dietpi-drive_manager" to setup mounts                                                                                                                                             
#----------------------------------------------------------------                                                                                                                               
# NETWORK                                                                                                                                                                                       
#----------------------------------------------------------------                                                                                                                               
                                                                                                                                                                                                
                                                                                                                                                                                                
#----------------------------------------------------------------                                                                                                                               
# TMPFS                                                                                                                                                                                         
#----------------------------------------------------------------                                                                                                                               
tmpfs /tmp tmpfs size=1024M,noatime,lazytime,nodev,nosuid,mode=1777                                                                                                                             
tmpfs /var/log tmpfs size=50M,noatime,lazytime,nodev,nosuid,mode=1777                                                                                                                           
                                                                                                                                                                                                
#----------------------------------------------------------------                                                                                                                               
# MISC: ecryptfs, vboxsf (VirtualBox shared folder), gluster, bind mounts                                                                                                                       
#----------------------------------------------------------------                                                                                                                               
                                                                                                                                                                                                
                                                                                                                                                                                                
#----------------------------------------------------------------                                                                                                                               
# SWAPFILE                                                                                                                                                                                      
#----------------------------------------------------------------                                                                                                                               
/var/swap none swap sw                                                                                                                                                                          
                                                                                                                                                                                                
#----------------------------------------------------------------                                                                                                                               
# PHYSICAL DRIVES                                                                                                                                                                               
#----------------------------------------------------------------                                                                                                                               
PARTUUID=907af7d0-02 / f2fs noatime,lazytime,rw 0 1                                                                                                                                             
PARTUUID=907af7d0-01 /boot vfat noatime,lazytime,rw 0 2  
```
```
root@BH-3B:~# blkid                                                                                                                                                                             
/dev/sda1: LABEL_FATBOOT="boot" LABEL="boot" UUID="4AD7-B4D5" TYPE="vfat" PARTUUID="907af7d0-01"                                                                                                
/dev/sda2: UUID="958c40d1-809d-4351-bf31-cd28b1a57350" TYPE="f2fs" PARTUUID="907af7d0-02" 
```

**Comment by Joulinar** [comment]
hmm the issue seems to be with `BootFS` and this is `vfat` file system. At least that's the messages on your screen print above. Can you try `dosfsck /dev/sda1`

@MichaIng and ideas

Next to this you still have `Under-voltage detected` messages

**Comment by Cjkeenan** [comment]
```
root@BH-3B:~# dosfsck /dev/sda1                                                                                                                                                                 
fsck.fat 4.1 (2017-01-24)                                                                                                                                                                       
0x41: Dirty bit is set. Fs was not properly unmounted and some data may be corrupt.                                                                                                             
1) Remove dirty bit                                                                                                                                                                             
2) No action                                                                                                                                                                                    
? 2                                                                                                                                                                                             
There are differences between boot sector and its backup.                                                                                                                                       
This is mostly harmless. Differences: (offset:original/backup)                                                                                                                                  
  65:01/00                                                                                                                                                                                      
1) Copy original to backup                                                                                                                                                                      
2) Copy backup to original                                                                                                                                                                      
3) No action                                                                                                                                                                                    
? 3                                                                                                                                                                                             
/dev/sda1: 311 files, 113448/516190 clusters  
```

The undervoltage messages are back because I switched back to the original power method since I would have to reconfigure my power strip a lot to fit that Canakit brick in. If it comes to that so be it and if you think that will help with troubleshooting I can do it, but it just is a lot easier with the USB port on the strip.

**Comment by Joulinar** [comment]
I know it's easier having the SSD connected via USB only but your are on a high risk of running an unstable system. Maybe your issues on BootFS are already a result of this.

But let's wait on @MichaIng Usually he has good ideas if it comes down to file system issues.

**Comment by MichaIng** [comment]
It is probably all related to the SSD powered via USB port. I am not 100% sure how the circuits are made on each RPi model, if USB device can have any effect on ARM voltage, but especially in case of peaks there is practically always an effect. So before investing time in any other investigations, first try to power the SSD via dedicated power supply and fix the under-voltage. Hard to not assume that failing external drive mounts, under-voltage warnings, and the external drive bring powered from RPi USB ports with 1.2A shared current are not related 😉. And of course random unmount and unstable voltage easily lead to file system errors. Of course it cannot be really said for a single error.

I just checked which dependencies the /boot mount actually ships, as I wasn't aware of any side of the parent mounts, which obviously was mounted fine (the root file system).
```
Requires=systemd-fsck@dev-disk-by\x2dpartuuid-907af7d0\x2d01.service
```
It seems that explicitly for FAT file systems, a file system check is mandatory at boot. So wile the above (dedicated power supply and fixing under-voltage warnings) stays true, fixing that file system error might (temporarily) solve your issue. Since the system boots, it should be safe to choose option `1)` in both cases, removing the dirty bit (mandatory) and copying the original boot sector to the backup. ... ah, verified something that I had in mind already: RPi does not need any boot sector, so it is totally irrelevant 😉.

**Comment by Cjkeenan** [comment]
Ok so what course of action should I take? My SSD enclosure does not have any external power option. I have not seen any enclosure that has external power, except those for 3.5" HDD drives. Like I said though, even with the Canakit brick, the undervoltage errors go away, but all of the other errors remain, is the damage already done and I need to do some sort of repair?

So I reconfigured to accommodate the Canakit brick and now the Voltage warning is gone, but even after clearing the dirty bit, and rebooting without power loss, same issues are present.
```
root@BH-3B:/boot# dmesg -l emerg,alert,crit,err                                                                                                                                                 
[    1.667748] vc_vchi_sm_init: failed to open VCHI service (-1)                                                                                                                                
[    1.667751] [vc_sm_connected_init]: failed to initialize shared memory service                                                                                                               
[  104.511092] vc_sm_cma_vchi_init: failed to open VCHI service (-1)                                                                                                                            
[  104.511111] [vc_sm_connected_init]: failed to initialize shared memory service                                                                                                               
[  104.817728] bcm2835_mmal_vchiq: Failed to open VCHI service connection (status=-1)  
```

**Comment by MichaIng** [comment]
Did you fix the difference in boot sector as well? In case run `dosfsck -a /dev/sda1` to have fixes applied automatically.
... but ah, actually having another look at your screenshot, the failing fsck is not the issue but another symptom. Waiting for the block device itself times out, hence neither can it be mounted, not can the fsck run.
A simple solution might be to give it some more time to load: Add `boot_delay=5` or even `boot_delay=10` to `/boot/config.txt` to give it some more time to be initialised.

Btw, the other kernel errors can be ignored. Those are expected when running the RPi with less than 32M GPU memory, where some GPU features are disabled (cut-down firmware files loaded). I already asked RPi devs if there is a way to prevent the attempt to load the vcsm device tree part as well as the kernel module, with cut-down firmware, since those are doomed to fail anyway and only produce error messages which might confuse or create worries, but it is not foreseen that the loaded firmware files can influence the device tree or loaded kernel modules. I really have to implement my personal workaround for this to disable both manually 😄.

Another thing: You have the serial/UART device disabled (`/dev/ttyS0`), but the serial console on that device enabled, obviously. So you might want to disable it, either via `dietpi-config` > `Advanced Options` > `Serial/UART` or manually via `systemctl disable serial-getty@ttyS0`.

**Comment by Cjkeenan** [comment]
> A simple solution might be to give it some more time to load: Add `boot_delay=5` or even `boot_delay=10` to `/boot/cmdline.txt` to give it some more time to be initialised.

Are you sure it is in the cmdline.txt? I found this [documentation](https://www.raspberrypi.org/documentation/configuration/config-txt/boot.md) that states it should be in the config.txt

> Another thing: You have the serial/UART device disabled (`/dev/ttyS0`), but the serial console on that device enabled, obviously.

Will do and will report back.

**Comment by Cjkeenan** [comment]
**Update:** No dice on adding a boot_delay=10 to the config.txt and disabling the serial console, same issues. Anything else you guys can think of or do you want to see any more logs? Is there any way to verify if that delay was indeed adhered to?

I found this on RPi's website in regards to [Mass Storage Booting](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bootmodes/msd.md):
_Known issues (not Pi 4)
The default timeout for checking bootable USB devices is 2 seconds. Some flash drives and hard disks power up too slowly. It is possible to extend this timeout to five seconds (add a new file timeout to the SD card), but note that some devices take even longer to respond._

Any ideas where to put that file? They reference a SD Card but that seems like a typo to me, or maybe with the timeout file it sets a different value to that OTP bit.

The weirdest thing to me is that it errors out and says it cant find that drive, presumably because it has not been mounted, but then goes into emergency mode, then I press CTRL+D to continue and it boots fine, so how does that make any sense?

**Comment by MichaIng** [comment]
<details>
<summary>Not really relevant in your case ;)</summary>
As far as I understand, you put a new (empty) file `timeout` onto the first partition of the SD card. The SD card is read/checked anyway, so it could see that file and then wait and see if a bootable USB drive is detected. What I don't understand in this regards is, that AFAIK the RPi boots from the SD card automatically if possible, also the docs state it:

> Make sure that you do not have an SD card inserted in the Pi, since if you do, it will boot from that first.

So then the SD cards must contain a FAT partition, likely on the same location as if an OS was installed, and contain that empty `timeout` file, but the SD card must not contain an OS aside of that, i.e. at least the `bootcode.bin` must not exist.

That docs are a scattered mess, heavily redundant, which makes it tedious to read. Apart of the `timeout` file, it is also possible to program this increased timeout via OTP bit: `program_usb_boot_timeout=1`

But before applying permanent changes to the OTP, I'd rule out everything else. The thing is that the RPi "boots" fine, bootloader, kernel, firmware from boot partition are read and loaded and root partition is mounted (?), it is only the Linux device connector that seems to be created very late somehow. By all I understand this timeout (2 vs 5 second) is only to allow more time for the very first step, the RPi ROM firmware to find a FAT partition with a `bootcode.bin` to start booting from. That stage has long been passed successfully.
</details>

Just to be sure, there is no SD card attached currently, is it?
And did you actually try to mount the `/boot` partition after CTRL+D boot? `mount /boot`
**Ah and I made a mistake above: `boot_delay=5`/`boot_delay=10` must not be added to `cmdline.txt `but to `config.txt`, sorry, I mixed that up. Fixed in above. Please try that first.**

**Comment by Cjkeenan** [comment]
Sorry, it appears we are on opposite sleep schedules, but anyway, there is no SD card attached and I did add `boot_delay=10` to the config.txt and it did not help.

```
root@BH-3B:/boot# mount /boot                                                                                                                                                                   
mount: /boot: /dev/sda1 already mounted on /boot. 
```
Also is it normal for the green LED be blinking at a 1s interval?

**Comment by Cjkeenan** [comment]
Not trying to push, and if you guys are busy I get it, but any other ideas of how to fix or at least troubleshoot?

**Comment by MichaIng** [comment]
The green LED beside the constantly lighting red led, right?
It should blink on disk writes, but you could actually check via `dietpi-led_control` what the trigger is and adjust it. So a constant 1s blink should actually not be the case. There are error codes, but AFAIK those only occur when boot fails completely: https://www.raspberrypi.org/documentation/configuration/led_blink_warnings.md

Is the error and solution you face still the same as in OP? Because `/boot` seems to have been successfully mounted according to your post 4 days ago?

Probably a fresh and complete `dmesg` and/or `journalctl` output can help to review everything that happens during boot.

**Comment by Cjkeenan** [comment]
Yes all of the same issues are present as the OP.

> `ietpi-led_control` what the trigger is and adjust it

I just reset all values to default and it has the same behavior, 1s blinks during and after boot

> Probably a fresh and complete `dmesg` and/or `journalctl` output can help to review everything that happens during boot.

[journalctl.txt](https://github.com/MichaIng/DietPi/files/5534386/journalctl.txt)
[dmesg.txt](https://github.com/MichaIng/DietPi/files/5534388/dmesg.txt)

**Comment by nonnorm** [comment]
Not sure, but 
```text
[    3.995637] usb 1-1.5: The driver for the USB controller dwc_otg_hcd does not support scatter-gather which is
[    3.995644] usb 1-1.5: required by the UAS driver. Please try an other USB controller if you wish to use UAS.
```
looks like the problem to me.  Doing a search, it seems like some people were able to resolve it by adding `max_usb_current=1` to `/boot/config.txt`

**Comment by Cjkeenan** [comment]
Issue still present even after including that max current definition in config.txt and I see the prior error is still there. I am guessing the controller that is having an issue is a hardware controller on the board and not just a software interface right?

New dmesg and journalctl files:
[dmesg2.txt](https://github.com/MichaIng/DietPi/files/5534791/dmesg2.txt)
[journalctl2.txt](https://github.com/MichaIng/DietPi/files/5534792/journalctl2.txt)

At this point I am beginning to think I may just upgrade to a RPi 4 to take advantage of the USB3. Would this actually help anything you think?

**Comment by MichaIng** [comment]
`max_usb_current` has not effect anymore, maximum USB current is available on all RPi models by default now.

The issue with `dwc_otg` driver and UAS is known and expected as well.
<details>
<summary>Excursus about the UAS story</summary>
Until RPi4, the default USB driver is `dwc-otg` which supports "FIQ" but no scatter-gather, which is required for UAS. There is an alternative mainstream diver `dwc2`, which supports scatter-gather, hence UAS, but no FIQ. Don't ask me about details and benefits of those technology, but the UAS kernel module was not even available until not very long ago. It has been build now since it includes a list of USB quicks which makes booting/mounting (don't remember the exact issue) of a very rare/special USB SATA bridge case possible, but actually real UAS was/is still not possible. I don't remember the details but it was definitely not a regular SATA drive attached via regular USB case/docking station, only reported by a single user. So now the UAS module is there, the kernel attempts to load it, which is known to be not possible with the default USB driver, UNTIL RPi 4.

The RPi 4 now loads the `dwc2` by default, which enables UAS support. Not sure why scather-gather+UAS is now seen more important than FIQ only on this model, the only thing I see now is regular reports with RPi 4 where booting or even mounting USB drive fails, since there are lots of drives (respectively contained firmware versions) available which do not support UAS and then fail. For this another USB quick needs to be added to blacklist UAS for that USB device. So to me it looks like actually enabling UAS now introduced a similar issue (from symptoms) for a large amount of users than what adding the kernel module in the first place was aimed to resolve for a single user only.

But UAS can be enabled on all RPi models by switching the USB driver: `echo 'dtoverlay=dwc2' >> /boot/config.txt`
And UAS can be disabled for all USB drives: `echo -e 'blacklist uas\nblacklist sg' > /etc/modprobe.d/disable_uas.conf`
The last one does not mute the error messages, but it saves ~50 KiB kernel memory space, for free on RPi <4 since they have no effect anyway.

Actually I wanted to ask the RPi devs about this for a long time, what the actual benefits of UAS are, if this really outweighs the issue that several users run into a failing USB drives without manual action. If there are benefits, why it is only enabled on RPi 4 by default, if the required `dwc2` driver could be a reasonable default on all RPi models, and otherwise if there are any downsides of blacklisting the uas and sg kernel modules (there are none, I'm doing this for a year on my RPi). Currently it looks wrong to me but I want to get some official info/opinions on this before applying an enhancement on DietPi 🙂.
</details>

Okay so we see the kernel waiting for the root device:
```
[    1.728302] Waiting for root device PARTUUID=907af7d0-02
```
The whole drive is quickly found, attached, all partitions seen correctly, fsck done on root (no dirty bit, hence continues immediately), then init system (systemd) is loaded:
```
[    5.100966]  sda: sda1 sda2
[    5.102845] sd 0:0:0:0: [sda] Attached SCSI disk
[    5.586670] F2FS-fs (sda2): recover fsync data on readonly fs
[    5.587276] F2FS-fs (sda2): Mounted with checkpoint version = 4e672421
[    5.587348] VFS: Mounted root (f2fs filesystem) readonly on device 8:2.
[    5.588788] devtmpfs: mounted
[    5.592356] Freeing unused kernel memory: 1024K
[    5.623938] Run /sbin/init as init process
```
And then we see the boot file system device not appearing before it times out after 90 seconds, which finally loads to emergency mode:
```
Nov 12 20:44:34 BH-3B systemd[1]: dev-disk-by\x2dpartuuid-907af7d0\x2d01.device: Job dev-disk-by\x2dpartuuid-907af7d0\x2d01.device/start timed out.
Nov 12 20:44:34 BH-3B systemd[1]: Timed out waiting for device /dev/disk/by-partuuid/907af7d0-01.
Nov 12 20:44:34 BH-3B systemd[1]: Dependency failed for File System Check on /dev/disk/by-partuuid/907af7d0-01.
Nov 12 20:44:34 BH-3B systemd[1]: Dependency failed for /boot.
Nov 12 20:44:34 BH-3B systemd[1]: Dependency failed for Local File Systems.
Nov 12 20:44:34 BH-3B systemd[1]: local-fs.target: Job local-fs.target/start failed with result 'dependency'.
Nov 12 20:44:34 BH-3B systemd[1]: local-fs.target: Triggering OnFailure= dependencies.
Nov 12 20:44:34 BH-3B systemd[1]: boot.mount: Job boot.mount/start failed with result 'dependency'.
Nov 12 20:44:34 BH-3B systemd[1]: systemd-fsck@dev-disk-by\x2dpartuuid-907af7d0\x2d01.service: Job systemd-fsck@dev-disk-by\x2dpartuuid-907af7d0\x2d01.service/start failed with result 'dependency'.
Nov 12 20:44:34 BH-3B systemd[1]: dev-disk-by\x2dpartuuid-907af7d0\x2d01.device: Job dev-disk-by\x2dpartuuid-907af7d0\x2d01.device/start failed with result 'timeout'.
```
Another fsck is done on the root fs (not the affected boot fs). I was wondering first about this:
```
Nov 12 20:44:45 BH-3B systemd-fsck[122]: Info: MKFS version
Nov 12 20:44:45 BH-3B systemd-fsck[122]:   "Linux version 5.8.0-1-amd64 (debian-kernel@lists.debian.org) (gcc-10 (Debian 10.2.0-6) 10.2.0, GNU ld (GNU Binutils for Debian) 2.35) #1 SMP Debian 5.8.7-1 (2020-09-05)"
```
But I think this just tells us about the host system which created the F2FS file system and this of course does not need to match the RPi.
Afterwards all continues fine, the previously missing boot partition device is there, fsck runs successfully on it, all as expected. So the big question is why the boot partition device is not appearing in the first place.

I don't think it can have an effect, but I recognised those two lines being nearly the last two before the timeout, about entropy pool initialisation:
```
[    6.911297] random: crng init done
[    6.911321] random: 7 urandom warning(s) missed due to ratelimiting
```
The entropy daemon starts much later:
```
Nov 12 20:45:00 BH-3B rng-tools[362]: Starting Hardware RNG entropy gatherer daemon: rngd.
Nov 12 20:45:00 BH-3B rngd[369]: rngd 2-unofficial-mt.14 starting up...
Nov 12 20:45:00 BH-3B systemd[1]: Started rng-tools.service.
```
There is a newer version that I actually wanted to pre-install on all RPi models, but one test on RPi 1 and Zero is outstanding:
```
apt install rng-tools5
apt purge rng-tools
```
However, it starts earlier but still after all local mounts have been done. Still, just to rule entropy out for sure, you could hit some random keyboard buttons during boot to create entropy. Ah and to better see what the current boot stage is, remove `quiet` from `/boot/cmdline.txt`.

And as well just to rule it out you could try to blacklist the UAS modules:
```
echo -e 'blacklist uas\nblacklist sg' > /etc/modprobe.d/disable_uas.conf
```
And if that has no effect, the other way round, enable UAS:
```
rm -f /etc/modprobe.d/disable_uas.conf
echo 'dtoverlay=dwc2' >> /boot/config.txt
```

From which DIetPi version were you upgrading, respectively, do you remember if it was still Linux 4.19? I would not wonder if some change in the kernel 5.4 is responsible. Actually, you could even try to downgrade. If this solved the issue, then we should forward to RPi devs:
```
cd /tmp
wget http://archive.raspberrypi.org/debian/pool/main/r/raspberrypi-firmware/raspberrypi-kernel_1.20200601-1_armhf.deb
wget http://archive.raspberrypi.org/debian/pool/main/r/raspberrypi-firmware/raspberrypi-bootloader_1.20200601-1_armhf.deb
wget http://archive.raspberrypi.org/debian/pool/main/r/raspberrypi-firmware/libraspberrypi0_1.20200601-1_armhf.deb
wget http://archive.raspberrypi.org/debian/pool/main/r/raspberrypi-firmware/libraspberrypi-bin_1.20200601-1_armhf.deb
dpkg -i --force-downgrade *raspberrypi*.deb
rm *.deb
```

The other two things are still:
- F2FS partition: Creates some work, but probably worth a test if a fresh DietPi image, flashed onto the drive, then copying/syncing the files from the current system (root + boot partitions) over (or remove old files first, copy new files afterwards, so that no obsolete files are left). That way it's assured that partition and file system meta data is exactly the same as on a fresh image. I cannot really imagine how it can have an effect, especially when boot partition is initially read fine (kernel, bootloader etc) and root partition mounts fine etc, and especially since it worked perfectly fine before, but just to rule it out.
- External power supply: Probably the new kernel 5.4 changed the way how power is drawn across the device. I know a dedicated power supply for a 2.5" case is quite uncommon, but for SBCs actually reasonable. Otherwise a 3.5" case/docking station usually can take a 2.5" drive as well.

**Comment by Cjkeenan** [comment]
Thank you so much for the detailed response, I will continue updating this comment as I go through the steps. One thing I did notice was that during one of the reboots, right before it shut off, I noticed "Failed to unmount /boot", not sure if that is relevant though. The most puzzling thing to me though is that prior to updating to 6.33, it booted fine to my knowledge, now the further that memory goes the more unsure, but I am like 90% sure I never had this issue.

> From which DIetPi version were you upgrading, respectively, do you remember if it was still Linux 4.19?

The initial build was the F2FS image you made here: https://github.com/MichaIng/DietPi/issues/606#issuecomment-695821494

> Still, just to rule entropy out for sure, you could hit some random keyboard buttons during boot to create entropy.

Did not seem to help

Files:
[journalctl.txt](https://github.com/MichaIng/DietPi/files/5539648/journalctl.txt)
[dmesg.txt](https://github.com/MichaIng/DietPi/files/5539650/dmesg.txt)

> And as well just to rule it out you could try to blacklist the UAS modules:

Did not seem to help

Files:
[dmesg_UASoff.txt](https://github.com/MichaIng/DietPi/files/5539661/dmesg_UASoff.txt)
[journalctl_UASoff.txt](https://github.com/MichaIng/DietPi/files/5539662/journalctl_UASoff.txt)

> And if that has no effect, the other way round, enable UAS:

Seemed to change something as it had a very different boot process than what has been happening and I left it for upwards of 5 minutes and it was stuck here:
![IMG_20201113_151211](https://user-images.githubusercontent.com/10240016/99129952-d8bc4680-25c3-11eb-8a32-ed74308241ea.jpg)

Files:
I could not generate due to a failed boot.

> Actually, you could even try to downgrade. If this solved the issue, then we should forward to RPi devs:

Did not seem to help (UAS disabled with rng_tools5)

Files:
[dmesg_kernelDown.txt](https://github.com/MichaIng/DietPi/files/5539751/dmesg_kernelDown.txt)
[journalctl_kernelDown.txt](https://github.com/MichaIng/DietPi/files/5539752/journalctl_kernelDown.txt)

**Comment by Cjkeenan** [comment]
I am very thankful for all of the troubleshooting steps you guys have provided but at this point should I just backup dietpi onto a flash drive, then reflash the F2FS image and then restore from backup or will that likely restore the bug as well? Like I said, while running everything seems fine, just in the event of a power outage, and I am not home, the RPi will not boot on its own and that is not great.

**Comment by Cjkeenan** [comment]
I am very thankful for all of the troubleshooting steps you guys have provided but at this point should I just backup dietpi onto a flash drive, then reflash the F2FS image and then restore from backup or will that likely restore the bug as well? Like I said, while running everything seems fine, just in the event of a power outage, and I am not home, the RPi will not boot on its own and that is not great.

**Comment by Joulinar** [comment]
Usually a `dietpi-backup` is a full backup. Probably the issue will be backed up as well.

**Comment by Cjkeenan** [comment]
So when I was looking into tips for making the RPi more reliable, I found this article and this specific tip. I wonder, do you think this is what is going on here, something with my clock is screwed up and it isn't allowing boot?

https://blog.ronnyvdb.net/2019/01/20/howto-make-a-raspberry-pi-truly-read-only-reliable-and-trouble-free/

step 4.4 and 4.5

**Comment by Joulinar** [comment]
usually system time is synchronized during boot. You can check it using `systemctl status systemd-timesyncd.service`

**Comment by MichaIng** [comment]
That guide is btw pretty outdated, as it handles things if there was no systemd. Most things do not apply on DietPi or are for the specific use case of the author (SSH tunnel) or standard/implemented in DietPi already.

Use `dietpi-config` > `Advanced Options` > `Time sync mode` > `4 : Daemon + Drift` to have `systemd-timesyncd` enabled permanent, like NTP or htpdate daemons would be. It starts slightly earlier on boot and of course keeps the accuracy higher compared to running it on hourly/daily basis, but otherwise there will be no significant difference. The basic problem that system clock is out of sync until the first network sync has been done, remains of course with a read-only file system.

Btw, instead of dealing with all the error messages when mounting the rootfs read-only, mounting an overlayfs that emulates R/W might be better. See here the instructions by @Joulinar: https://dietpi.com/phpbb/viewtopic.php?p=24920#p24920

**Comment by Cjkeenan** [comment]
No dice on it being the Time Sync Mode. For reference, I did not follow that guide, but rather I was just looking around for similar symptoms and came across that. Do you guys have any inclination as to what is going on? It seems rather odd to me that it times out waiting for the device, and then after forcing it via CTRL+D, it magically mounts and boots fine no?

**Comment by MichaIng** [comment]
Can you please show your `cmdline.txt`: `cat /boot/cmdline.txt`
It has the `rootfstype` boot parameter, if that is set correctly to "f2fs", probably removing that option helps.

Ah, and can you show `lsmod`, to see if F2FS is built into the kernel or loaded as module. In the letter case, it would make sense that it cannot be mounted at boot without initramfs (default on RPi). Would be strange then that it works after CTRL+D, but let's simply rule it out 😉.

**Comment by Cjkeenan** [comment]
> It has the `rootfstype` boot parameter, if that is set correctly to "f2fs", probably removing that option helps.

```
dietpi@BH-3B:~$ cat /boot/cmdline.txt                                                                                                                                                           
 console=tty1 root=PARTUUID=907af7d0-02 rootfstype=f2fs elevator=deadline fsck.repair=yes rootwait net.ifnames=0     
 ```
Remove the option altogether or should I put something else there? FYI, removing altogether did not seem to help.

> Ah, and can you show lsmod, to see if F2FS is built into the kernel or loaded as module.

```
dietpi@BH-3B:~$ lsmod                                                                                                                                                                           
Module                  Size  Used by                                                                                                                                                           
xt_nat                 16384  0                                                                                                                                                                 
xt_tcpudp              16384  0                                                                                                                                                                 
veth                   28672  0                                                                                                                                                                 
xt_conntrack           16384  2                                                                                                                                                                 
nf_conntrack_netlink    45056  0                                                                                                                                                                
xt_addrtype            16384  2                                                                                                                                                                 
br_netfilter           28672  0                                                                                                                                                                 
bridge                139264  1 br_netfilter                                                                                                                                                    
stp                    16384  1 bridge                                                                                                                                                          
llc                    16384  2 bridge,stp                                                                                                                                                      
wireguard             131072  0                                                                                                                                                                 
ip6_udp_tunnel         16384  1 wireguard                                                                                                                                                       
udp_tunnel             16384  1 wireguard                                                                                                                                                       
overlay               102400  0                                                                                                                                                                 
sg                     28672  0                                                                                                                                                                 
joydev                 20480  0                                                                                                                                                                 
uas                    24576  0                                                                                                                                                                 
bcm2835_isp            32768  0                                                                                                                                                                 
bcm2835_mmal_vchiq     28672  1 bcm2835_isp                                                                                                                                                     
vc_sm_cma              32768  2 bcm2835_isp,bcm2835_mmal_vchiq                                                                                                                                  
videobuf2_dma_contig    20480  1 bcm2835_isp                                                                                                                                                    
videobuf2_memops       16384  1 videobuf2_dma_contig                                                                                                                                            
videobuf2_v4l2         28672  1 bcm2835_isp                                                                                                                                                     
videobuf2_common       57344  2 bcm2835_isp,videobuf2_v4l2                                                                                                                                      
raspberrypi_hwmon      16384  0                                                                                                                                                                 
videodev              237568  3 bcm2835_isp,videobuf2_common,videobuf2_v4l2                                                                                                                     
mc                     40960  4 bcm2835_isp,videobuf2_common,videodev,videobuf2_v4l2                                                                                                            
uio_pdrv_genirq        16384  0                                                                                                                                                                 
uio                    20480  1 uio_pdrv_genirq                                                                                                                                                 
fixed                  16384  0                                                                                                                                                                 
nft_counter            16384  24                                                                                                                                                                
nft_chain_nat          16384  4                                                                                                                                                                 
xt_MASQUERADE          16384  3                                                                                                                                                                 
nf_nat                 49152  3 xt_nat,xt_MASQUERADE,nft_chain_nat                                                                                                                              
nf_conntrack          135168  5 xt_nat,xt_MASQUERADE,nf_conntrack_netlink,xt_conntrack,nf_nat                                                                                                   
nf_defrag_ipv6         20480  1 nf_conntrack                                                                                                                                                    
nf_defrag_ipv4         16384  1 nf_conntrack                                                                                                                                                    
xt_comment             16384  0                                                                                                                                                                 
nft_compat             20480  7                                                                                                                                                                 
nf_tables             139264  82 nft_compat,nft_chain_nat,nft_counter                                                                                                                           
nfnetlink              16384  4 nft_compat,nf_conntrack_netlink,nf_tables                                                                                                                       
ip_tables              28672  0                                                                                                                                                                 
x_tables               32768  8 xt_comment,xt_nat,ip_tables,nft_compat,xt_tcpudp,xt_MASQUERADE,xt_addrtype,xt_conntrack                                                                         
ipv6                  458752  40 br_netfilter,wireguard,bridge
```

**Comment by MichaIng** [comment]
Okay, also here I see it being as kernel builtin:
```
2021-01-11 16:39:53 root@micha:~# modinfo f2fs
name:           f2fs
filename:       (builtin)
license:        GPL
file:           fs/f2fs/f2fs
description:    Flash Friendly File System
author:         Samsung Electronics's Praesto Team
alias:          fs-f2fs
```
This means no initramfs required. Okay I'm out of ideas what the reason could be, but it most likely has to do with F2FS. I'm not finding time to test myself soon, probably search machines can help to check for similar issues or requirements when booting an RPi from F2FS.

**Comment by Cjkeenan** [comment]
So since this issue is still present and cannot be easily fixed it appears, I am just going to chalk it up to a fluke corruption.

What is the best method for installing a fresh copy of a f2fs install? Should I install raspbian as f2fs and then do the script to make my own distribution or is there an easier way?

**Comment by MichaIng** [comment]
Since there is no F2FS image for Raspbian/Raspberry Pi OS (or is it?), you could install DietPi as F2FS just as you would do with a fresh Raspberry Pi OS image, e.g.:
- Flash a fresh DietPi image onto a new SD card and mount it to another Linux system.
- Copy the content of the 2nd partition somewhere else.
- Delete and re-create the second partition with full size, and format it with F2FS file system.
- Copy the content back in.
- Edit `cmdline.txt` on the 1st partition and `/etc/fstab` on the second partition so that file system type and PARTUUID of the root partition match our new 2nd partition.

That should be it. Since the DietPi fs_partition_resize script does not support F2FS yet, it'll fail, but no issue since partition and file system is extended already.

**Comment by Cjkeenan** [comment]
> Since there is no F2FS image for Raspbian/Raspberry Pi OS (or is it?)

Officially? No, but I found this utility (https://github.com/FrangaL/rpi-img-builder) which has a bunch of options to pre-setup an image including selecting a filesystem, so my plan was to build a lite Raspberry pi OS image with f2fs then upgrade to dietpi via your script.

I wonder if that developer would be interested/able to provide dietpi as another option for OS, which would be very cool.

**Comment by MichaIng** [comment]
Ah yes, that'll work, the script looks good 👍.

> I wonder if that developer would be interested/able to provide dietpi as another option for OS, which would be very cool.

The script installs a very basic Debian or Raspbian system, which moreless matches with the DietPi base system (repository, packages, even most of the applied system configs match to how we ship it). But a DietPi image has configs applied via our scripts and first boot services enabled in a way that at first boot `dietpi.txt` choices are applied, `dietpi-config` (and other tools) detect and apply configs correctly etc. It would be not reasonable to implement and especially maintain this in a separate script when one can simply apply it via our maintained script in a second step: https://github.com/MichaIng/DietPi/blob/dev/PREP_SYSTEM_FOR_DIETPI.sh
Most of the code lines are pre-image specific cleanup and board-specific setup things, so only a minor part would need to be ported when using a debootstraped Raspbian, but already for downloading and installing the `dietpi-*` scripts and enabling the first boot services it wouldn't make sense to double the coding effort, IMO.

DietPi simply is not only a minimal image. Such a debootstraped Raspbian (when choosing "slim" in the script) is the naked minimal image, which in turn excludes the DietPi tools to take users by the hand at first boot and for software installs 😉.

**Comment by MichaIng** [comment]
I mark the issue as closed. Feel free to reopen if required..

**Comment by aakerbeere** [comment]
> At this point I am beginning to think I may just upgrade to a RPi 4 to take advantage of the USB3. Would this actually help anything you think?

My RPi4 is suffering from the same issue ([see here](https://github.com/raspberrypi/linux/issues/3108#issuecomment-887484610)).
Actually i'm back on ext4 for the "root"- Partition, but would prefere f2fs.

cheers

**Comment by MichaIng** [comment]
Can you please share in detail all steps you did to migrate that DietPi system onto the F2FS filesystem?

**Comment by aakerbeere** [comment]
My equipement is not DietPi but Pi4.
It did boot from USB Samsung Portable SSD T5 having f2fs for "/root"-partition and it did not boot anymore after the nextcloud instance having done syncing some additional data.

Plugged the drive to notebook, rsynced the data, gparted the partition to ext4, rsynced data back, edited fstab and cmdline.txt and on it goes.

Would do exactly the same vice versa (and have done it before) to revert running system to f2fs.

**Comment by MichaIng** [comment]
> My equipement is not DietPi but Pi4.

What do you mean by this? DietPi is an operating system, Pi4 (= Raspberry Pi 4) is an SBC. If you do not run DietPi on your Pi, please check for support on the operating system you use 😉.

**Comment by aakerbeere** [comment]
Sry
I didn't know DietPi. I only wanted to indicate to Cjkeenan he could get the same issue with a Pi4, because he was asking if it would bee an idea to upgrade the HW to Pi4.
I'm running RPiOS 64bit Buster and get the same issue as he did/does? with DietPi.

**Comment by MichaIng** [comment]
Well, we already tested DietPi successfully on F2FS, so it's at least nothing with the kernel, firmware or bootloader parts. I guess it is something with the Raspberry Pi OS specific setup scripts then.
